lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -771,6 +761,92 @@ mkTipSyncClient tr np localTxSubmissionQ onTipUpdate onPParamsUpdate onInterpret
         NodeToClientV_3
 
 
+-- Reward Account Balances
+
+-- | Monitors values for keys, and allows clients to @query@ them.
+--
+-- Designed to be used for observing reward balances, where we want to query the
+-- balances of /all/ the wallets' accounts on tip change, and allow wallet
+-- workers to @query@ the value.
+--
+-- One could imagine replacing @query@ with a push-based approach.
+data Observer m key value = Observer
+    { startObserving :: key -> m ()
+    , stopObserving :: key -> m ()
+    , query :: key -> m (Maybe value)
+    }
+
+newRewardBalanceFetcher
+    :: forall sc. Tracer IO (NetworkLayerLog sc)
+    -> W.GenesisParameters
+    -- ^ Used to convert tips for logging
+    -> TQueue IO (LocalStateQueryCmd (CardanoBlock sc) IO)
+    -> IO ( Observer IO W.ChimericAccount W.Coin
+          , Tip (CardanoBlock sc) -> IO ()
+            -- Call on tip-change to refresh
+          )
+newRewardBalanceFetcher tr gp queryRewardQ = do
+    cacheVar <- atomically $ newTVar Map.empty
+    toBeObservedVar <- atomically $ newTVar Set.empty
+    return (observer cacheVar toBeObservedVar, refresh cacheVar toBeObservedVar)
+  where
+    observer
+        :: TVar IO (Map W.ChimericAccount W.Coin)
+        -> TVar IO (Set W.ChimericAccount)
+        -> Observer IO W.ChimericAccount W.Coin
+    observer cacheVar toBeObservedVar =
+        Observer
+            { startObserving = \k ->
+                atomically $ do
+                    modifyTVar' toBeObservedVar (Set.insert k)
+            , stopObserving = \k ->
+                atomically $ do
+                    modifyTVar' toBeObservedVar (Set.delete k)
+                    modifyTVar' cacheVar (Map.delete k)
+            , query = \k -> do
+                m <- atomically (readTVar cacheVar)
+                return $ Map.lookup k m
+            }
+
+    refresh
+        :: TVar IO (Map W.ChimericAccount W.Coin)
+        -> TVar IO (Set W.ChimericAccount)
+        -> Tip (CardanoBlock sc)
+        -> IO ()
+    refresh cacheVar toBeObservedVar tip = do
```

Comment:
I find it surprising that `refresh` is not part of the `Observer` interfaces. The function clearly has a privileged access to both TVar and make strong assumption about how they are manipulated. So to me, it seems like it belong to the same abstraction and having it as a separate function is a bit confusing.

Reply:
To me, `Observer` is concerned about _subscribing to_ and _accessing_ values. It theoretically be passed around and escape the `Network` module. 

`refresh` is coupled to the _creation_ of the `Observer`. It's more like an implementation detail, specific to the `Network` module.

If say the wallet registry were to have access to an `Observer`, it should not be allowed to manually call `refresh`.